### PR TITLE
Updates for the upcoming new Insights outputs

### PR DESCRIPTION
**Pull request by autarch**

### CODE REVIEW

File: src/main/java/com/maxmind/minfraud/response/Device.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+public final class Device {
+    private final String id;
+
+    public Device(
+            @JsonProperty("id") String id
+    ) {
+        this.id = id;
+    }
+
+    public Device() {
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/Device.java]: Do we need these empty constructors for anything (also in the Email class)?

File: src/main/java/com/maxmind/minfraud/response/InsightsResponse.java [hunk 2]
```diff
@@ -72,6 +94,7 @@ public String toString() {
         sb.append(", warnings=").append(this.warnings);
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/InsightsResponse.java]: You didn't add `device` to the `toString()`. That said, for GeoIP2, I replaced the `toString()` methods with [output created from the JSON](https://github.com/maxmind/GeoIP2-java/commit/7706e126a6768a76e76191f36e18f454a8ce329c). We could do that for these model classes too. `toString()` is used primarily as a debugging aid for programmers.

File: src/main/java/com/maxmind/minfraud/response/CreditCard.java [hunk 3]
```diff
@@ -67,14 +81,49 @@ public Boolean isPrepaid() {
...
     }
+
+    /**
+     * The enumerated credit card types
+     */
+    public enum Type {
+        BLANK,
+        CHARGE,
+        CREDIT,
+        DEBIT;
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/CreditCard.java]: If we are going to have an enum here, we should have a test that ensure a new value in the response does not break it.

File: src/main/java/com/maxmind/minfraud/response/Device.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        this(null);
+    }
+
+    /**
+     * @return The device id.
+     */
+    public String getId() {
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/Device.java]: I wonder if we should make this an actual `UUID` object (as well as minFraud ID). That's essentially what I did in the .NET reader.

File: src/main/java/com/maxmind/minfraud/response/Email.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+     * @return True if the email address is for a free email service provider.
+     */
+    public Boolean isFree() {
+        return isFree;
+    }
+
+    /**
+     * @return True if the email address is associated with fraud.
+     */
+    public Boolean isHighRisk() {
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/Email.java]: I don't think these will be serialized correctly with the `toJson()` method as they are missing the `@JsonProperty("is_high_risk")`, etc., annotations.

I _think_ this is tested, but perhaps the new inputs did not end up in that test.

File: src/main/java/com/maxmind/minfraud/response/Device.java (hunk 1 above)
Comment (autarch) [inline src/main/java/com/maxmind/minfraud/response/Device.java]: I was mostly cutting and pasting here. Can I just move the `@JsonProperty` to the attribute declaration instead?

File: src/main/java/com/maxmind/minfraud/response/InsightsResponse.java (hunk 2 above)
Comment (autarch) [inline src/main/java/com/maxmind/minfraud/response/InsightsResponse.java]: That seems like it should be a separate PR, but I agree with the idea.

File: src/main/java/com/maxmind/minfraud/response/Device.java (hunk 1 above)
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/Device.java]: I'd leave it on the constructor parameter.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,19 @@
...
+
+* Added support for new minFraud Insights outputs. These are:
+    * `/credit_card/brand`
+    * `/credit_card/type`
+    * `/device/id`
+    * `/email/is_free`
+    * `/emai/is_high_risk`
+* The `Warning.input` attribute has been replaced by `Warning.inputPointer`,
+  which is a JSON Pointer rather than array.
+* The `ScoreResponse.id` and `InsightsResponse.id` attributes are now `UUID`
```
Comment (oschwald) [inline CHANGELOG.md]: These are private. You should probably rephrase this in terms of the getter. Same with the warning attributes.

File: src/main/java/com/maxmind/minfraud/response/Email.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
+package com.maxmind.minfraud.response;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * This class contains minFraud response data related to the email address.
+ */
+public final class Email {
+    @JsonProperty("is_free")
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/Email.java]: You don't need these. Jackson can't even access the attributes directly as we turned off overriding of access modifiers.

File: src/main/java/com/maxmind/minfraud/response/ScoreResponse.java [hunk 8]
```diff
@@ -26,7 +27,7 @@ public ScoreResponse(
             @JsonProperty("warnings") List<Warning> warnings
     ) {
         this.creditsRemaining = creditsRemaining;
-        this.id = id;
+        this.id = id == null ? null : UUID.fromString(id);
```
Comment (oschwald) [inline src/main/java/com/maxmind/minfraud/response/ScoreResponse.java]: Did you try making the parameter a UUID? It is possible that Jackson will just do the right thing.
